Logic

IgorPgrm opened the pull request.

IgorPgrm committed:
Scaffold Answer, переписан контроллер, добавлен в роут

IgorPgrm committed:
Partial _answer, виды

IgorPgrm committed:
Удалены ненужные стили

IgorPgrm committed:
-Удалена таблица user_tests
+Добавлена таблица test_passages
изменены ассоциации

IgorPgrm committed:
+TestPassages controller, виды, старт теста, отрисовка результата

IgorPgrm committed:
Назначение следующего вопроса, обратный вызов

IgorPgrm committed:
+Назначение следующего вопроса колбеком
Вид результата пройденного теста: Процент, цвет и уведомление на странице
Добавлена кнопка на список тестов

IgorPgrm committed:
+Хелпер, вычисляет номер текущего вопроса из общего кол-ва вопросов (0/4)
+Вид, отображение текущего и общего кол-ва вопросов

BubuntuClu commented on app/helpers/test_passages_helper.rb:
```diff
@@ -0,0 +1,9 @@
+    current_question_number = 0
+    total_question_ids = @test_passage.test.questions.ids
+    current_question_id = @test_passage.current_question.id
+    total_question_ids.each.with_index(1){ |x, index| current_question_number = index if x == current_question_id}
```
нахождение текущего номера вопроса сделай методом модели и находи его максимально БДшным способом. перебирать каждый вопрос не надо. найди общее количество тестов и используя текущий вопрос ты можешь понять сколько осталось

BubuntuClu commented on app/models/user.rb:
```diff
@@ -1,14 +1,18 @@
 
   def list_of_tests(level = 0)
-    Test.joins(:user_tests).where('user_id = ? and level= ?', id, level)
+    Test.joins(:test_passages).where('user_id = ? and level= ?', id, level)
```
`tests.where(level: level)`

BubuntuClu commented on app/views/test_passages/result.html.erb:
```diff
@@ -0,0 +1,29 @@
+<h1>The <%= @test_passage.test.title %> Test was Completed!</h1>
+<p><h2>
+  <% if @test_passage.percent >= 85 %>
```
что такое 85? без контекста это просто магическое число. надо убрать его в модель и обозвать понятно.
так же само это сравнение можно поместить в метод success?

BubuntuClu commented on db/migrate/20201217182644_drop_table_user_tests.rb:
```diff
@@ -0,0 +1,5 @@
+class DropTableUserTests < ActiveRecord::Migration[6.0]
+  def up
```
если есть up то должен быть down

BubuntuClu commented on db/migrate/20201217181901_create_test_passages.rb:
```diff
@@ -0,0 +1,15 @@
+
+      t.timestamps
+    end
+    add_index :test_passages, :users
```
индекс создается и так при `references` команде. нужно указать внешний ключ. задай его сразу внутри блока создания таблицы, а не за ним

BubuntuClu commented on db/schema.rb:
```diff
@@ -35,6 +35,21 @@
+    t.integer "correct_questions", default: 0
+    t.datetime "created_at", precision: 6, null: false
+    t.datetime "updated_at", precision: 6, null: false
+    t.index "\"current_questions\"", name: "index_test_passages_on_current_questions"
```
какие-то у тебя тут странные махинации в именах, не находишь?) удали эти индексы, которые ты создал руками. они какие-то кривые. нужно добавить внешний ключ

IgorPgrm committed:
Исправления
Нахождение номера текущего вопроса в модели
Проверка прохождения теста %
Добавлены внешние ключи в test_passages

IgorPgrm committed:
Исправлена миграция
Изменен seed (так как требует поле author_id)

BubuntuClu commented on app/models/question.rb:
```diff
@@ -7,6 +7,10 @@ class Question < ApplicationRecord
   validates :body, presence: true
   validate :answers_counter, on: :create
 
+  def self.remainder_question_count(question)
```
вопрос знает информацию только о самом себе, о других вопросах он не осведомлен. этот метод стоит поместить в test_passage

BubuntuClu commented on app/models/user.rb:
```diff
@@ -9,7 +9,7 @@ class User < ApplicationRecord
 
   def list_of_tests(level = 0)
-    Test.joins(:test_passages).where('user_id = ? and level= ?', id, level)
+    Tests.where(level: level)
```
невалидный руби код

BubuntuClu replied:
ты проверял этот код в консоле?

IgorPgrm committed:
Исправления

BubuntuClu commented on app/models/test_passage.rb:
```diff
@@ -24,13 +24,17 @@ def percent
 
   def current_question_number
-    self.test.questions.count - Question.remainder_question_count(self.current_question)
+    self.test.questions.count - self.remainder_question_count(self.current_question)
```
self в большинстве случаев можно опускать
